### PR TITLE
feat(autocomplete): Add autocomplete page

### PR DESCRIPTION
**Pull request by svemat01**
**Please describe the changes this PR makes and why it should be merged:**
This pull request adds a new page for autocomplete and roughly goes thru how to use them.

Slash command builder instructions are nonexistent as https://github.com/discordjs/builders/pull/53 is yet to be merged. Merge can be set on hold until the above-mentioned PR is merged if wanted.

## **For people looking for the guide. It's W.I.P but can be viewed [here](https://deploy-preview-943--discordjs-guide.netlify.app/interactions/autocomplete.html):**
https://deploy-preview-943--discordjs-guide.netlify.app/interactions/autocomplete.html

### CODE REVIEW

**Comment by lucemans** [comment]
LGTM!

File: guide/interactions/autocomplete.md
Comment (GodderE2D) [inline guide/interactions/autocomplete.md:15]:
```suggestion
To use autocomplete with your command options, you have to tell Discord to enable it.
```

File: guide/interactions/autocomplete.md
Comment (GodderE2D) [inline guide/interactions/autocomplete.md:18]:
```suggestion
### Slash command builder
```

File: guide/interactions/autocomplete.md
Comment (GodderE2D) [inline guide/interactions/autocomplete.md:22]:
```suggestion
### Custom system
```

File: guide/interactions/autocomplete.md
Comment (GodderE2D) [inline guide/interactions/autocomplete.md:58]:
```suggestion
Using `AutocompleteInteraction#respond()` you submit an array of ApplicationCommandOptionChoice objects. Passing an empty array will show "No options match your search" in Discord.
```

**Comment by Mysterious-Dev** [comment]
LGTM

File: guide/interactions/autocomplete.md
Comment (svemat01) [reply guide/interactions/autocomplete.md:16]: Yes but the builder doesn't yet support using autocomplete and thus why I decided on covering both.

File: guide/interactions/autocomplete.md
Comment (svemat01) [reply guide/interactions/autocomplete.md:58]: According to the ``CONTRIBUTING.md`` we should use "You"/"your" instead of "we"/"our" when explaining things in the guide.

Comment (Megatank58) [reply guide/interactions/autocomplete.md:58]: Oh my bad i readed it wrong 😅

Comment (svemat01) [reply guide/interactions/autocomplete.md:58]: Ah, No problem.
Should I reject your suggestions containing ``we`` and manually add back the other parts?

Comment (Megatank58) [reply guide/interactions/autocomplete.md:58]: I'll edit them to have `you/your` instead of `we` :grin:

File: guide/interactions/autocomplete.md
Comment (svemat01) [reply guide/interactions/autocomplete.md:62]: It looks better but couldn't this confuse people to use ``Array.filter()`` instead of ``myArray.Filter()``?

File: guide/interactions/autocomplete.md
Comment (svemat01) [reply guide/interactions/autocomplete.md:64]: According to the ``CONTRIBUTING.md`` references to class names should be capitalized, but remain outside of backticks.

File: guide/interactions/autocomplete.md
Comment (Megatank58) [reply guide/interactions/autocomplete.md:62]: `Array` is written as a class here, but that would be better as the variable here is `myArray`

Comment (svemat01) [reply guide/interactions/autocomplete.md:62]: Should it then be ``Array.filter()`` or ``Array#filter()``?

File: guide/interactions/autocomplete.md
Comment (Megatank58) [reply guide/interactions/autocomplete.md:64]: The select menu and button pages don't follow this though :thinking:

File: guide/interactions/autocomplete.md
Comment (Megatank58) [reply guide/interactions/autocomplete.md:62]: The latter of course

File: guide/interactions/autocomplete.md
Comment (svemat01) [reply guide/interactions/autocomplete.md:64]: I looked thru that page while making this page and there's a lot of points off the ``CONTRIBUTING.md`` that aren't being followed in the buttons and select menu guides. So It's a confusing situation, do we do like the other guides or do we follow the guidelines.

Comment (Megatank58) [reply guide/interactions/autocomplete.md:64]: New pages should follow the guidelines as they are in CONTRIBUTING.md, unsure about the rest but here it shouldn't be in backticks as you mentioned

File: guide/interactions/autocomplete.md
Comment (svemat01) [reply guide/interactions/autocomplete.md:55]: We should probably remove the backticks around `AutocompleteInteraction` as well.

File: guide/interactions/autocomplete.md
Comment (svemat01) [reply guide/interactions/autocomplete.md:58]: Remove backticks around `ApplicationCommandOptionChoice`

File: guide/interactions/autocomplete.md
Comment (svemat01) [reply guide/interactions/autocomplete.md:62]: Do you edit it?

File: guide/interactions/autocomplete.md
Comment (svemat01) [reply guide/interactions/autocomplete.md:64]: Same thing here with replacing ``Array.map()`` with ``Array#map()``

Comment (Megatank58) [reply guide/interactions/autocomplete.md:64]: Missed that 😅

File: guide/interactions/autocomplete.md
Comment (svemat01) [reply guide/interactions/autocomplete.md:58]: user in ``for the User`` should perhaps be lowercase?

File: guide/interactions/autocomplete.md
Comment (LukeHankey) [inline guide/interactions/autocomplete.md:35]: For this custom system, the `type: STRING` only works for when the client is logged in and this data is passed to `<Guild>.commands?.create()` (or other methods that take ApplicationCommandData). As all of this guide uses REST to register commands, shouldn't this be changed to the corresponding integer value `type: 3`?

Comment (svemat01) [reply guide/interactions/autocomplete.md:35]: I saw your discussion in discord, and your suggestion has a really good point, and using `type: 3` works with both and should thus be prefered. However, I'd suggest writing it as `type: 3, // String` to clarify.

Comment (LukeHankey) [reply guide/interactions/autocomplete.md:35]: I think that would work. Is it just the string option that is able to use autocomplete, or are there others too?

Comment (svemat01) [reply guide/interactions/autocomplete.md:35]: I think it's Integers as well, not 100% sure tho. And to clarify, my suggestion was just to add a comment at the end of `type: 3` to indicate that 3 means string.

Comment (LukeHankey) [reply guide/interactions/autocomplete.md:35]: Just tested that out. `integer` and `number` types can be used for autocomplete, 4 and 10 respectively. The others are basically autocompleted options.

File: guide/interactions/autocomplete.md
Comment (WiseDevHelper) [inline guide/interactions/autocomplete.md:44]:
```suggestion
To receive an <DocsLink path="class/AutocompleteInteraction" />, you can listen to the `interactionCreate` event and use the <DocsLink path="class/Interaction?scrollTo=isAutocomplete" /> method to make sure you only receive autocomplete interactions:
```

File: guide/interactions/autocomplete.md
Comment (WiseDevHelper) [inline guide/interactions/autocomplete.md:55]:
```suggestion
The <DocsLink path="class/AutocompleteInteraction" /> class provides the <DocsLink path="class/AutocompleteInteraction?scrollTo=respond" /> method for responding to the interaction.
```

File: guide/interactions/autocomplete.md
Comment (WiseDevHelper) [inline guide/interactions/autocomplete.md:58]:
```suggestion
Using <DocsLink path="class/AutocompleteInteraction?scrollTo=respond" /> you submit an array of <DocsLink path="typedef/ApplicationCommandOptionChoice" /> objects. Passing an empty array will show "No options match your search" for the user.
```

File: guide/interactions/autocomplete.md
Comment (WiseDevHelper) [inline guide/interactions/autocomplete.md:60]:
```suggestion
<DocsLink path="class/CommandInteractionOptionResolver?scrollTo=getFocused" /> returns the currently focused option's value. This value is used to filter the choices presented.
```

File: guide/interactions/autocomplete.md
Comment (WiseDevHelper) [inline guide/interactions/autocomplete.md:64]:
```suggestion
Using `Array#map()`, you can transform the array into an array of <DocsLink path="typedef/ApplicationCommandOptionChoice" /> objects.
```

or

```suggestion
Using `Array#map()`, you can transform the array into an array of `ApplicationCommandOptionChoice` objects.
```

File: guide/interactions/autocomplete.md
Comment (WiseDevHelper) [inline guide/interactions/autocomplete.md:85]: Maybe
```suggestion
To distinguish between multiple options you can pass `true` into <DocsLink path="class/CommandInteractionOptionResolver?scrollTo=getFocused" /> which now returns the full focused object instead of just the value. This is used to get the name of the focused option.
```

File: guide/interactions/autocomplete.md
Comment (WiseDevHelper) [inline guide/interactions/autocomplete.md:3]: - Slash command options provide the feature of autocomplete.
Maybe this sentence could be rephrased? And maybe tell the reader what Autocomplete is and why and where they should use it.

```suggestion
Slash command options provide the feature of autocomplete. In this section, we'll cover how to add autocomplete support to your command and how to respond to it.
```

**Review by WiseDevHelper — CHANGES_REQUESTED** [review]
LGTM, Just adding more refs to the docs.

File: guide/interactions/autocomplete.md
Comment (svemat01) [reply guide/interactions/autocomplete.md:35]: Please change your `type: 3,` to `type: 3, // String` before I can commit it.

File: guide/interactions/autocomplete.md
Comment (svemat01) [reply guide/interactions/autocomplete.md:58]: Could you add `can` after `<DocsLink path="class/AutocompleteInteraction?scrollTo=respond" /> you` so it ends up as
```md
<DocsLink path="class/AutocompleteInteraction?scrollTo=respond" /> you can submit
```

File: guide/interactions/autocomplete.md
Comment (LukeHankey) [reply guide/interactions/autocomplete.md:35]: > Please change your `type: 3,` to `type: 3, // String` before I can commit it.

👍

File: guide/interactions/autocomplete.md
Comment (adrifcastr) [inline guide/interactions/autocomplete.md:11]:
```suggestion
```
This seems irrelevant

File: guide/interactions/autocomplete.md
Comment (adrifcastr) [inline guide/interactions/autocomplete.md:3]: This feels run-on and inorganic, `of autocomplete` doesn't seem to be proper English, `command` should probably be plural or replaced with `options.`

File: guide/interactions/autocomplete.md
Comment (adrifcastr) [inline guide/interactions/autocomplete.md:16]:
```suggestion
```
This sounds weird. Should be rephrased into something along the lines of a more mature explanation of what is happening here.

File: guide/interactions/autocomplete.md
Comment (adrifcastr) [inline guide/interactions/autocomplete.md:20]:
```suggestion
```
If the section doesn't exist, there's no point in adding it in as a waiting ghost title.

File: guide/interactions/autocomplete.md
Comment (adrifcastr) [inline guide/interactions/autocomplete.md:24]:
```suggestion
```
Calling this a "custom system" feels out of place, what defines custom? This just an object with a property that holds a value, this can just be straightforward used as example.

File: guide/interactions/autocomplete.md
Comment (adrifcastr) [inline guide/interactions/autocomplete.md:55]:
```suggestion
The <DocsLink path="class/AutocompleteInteraction" /> class provides the <DocsLink path="class/AutocompleteInteraction?scrollTo=respond" /> method to send a response.
```
Reads better, less repetition.

File: guide/interactions/autocomplete.md
Comment (adrifcastr) [inline guide/interactions/autocomplete.md:57]:
```suggestion
### Sending results
```
same as above

File: guide/interactions/autocomplete.md
Comment (adrifcastr) [inline guide/interactions/autocomplete.md:58]:
```suggestion
Using `AutocompleteInteraction#respond()` you submit an array of ApplicationCommandOptionChoice objects. Passing an empty array will display "No options match your search" within the Discord client.
```
This needs to be rephrased to feel more organic and professional, `in Discord` again sounds how a child would describe it.

File: guide/interactions/autocomplete.md
Comment (adrifcastr) [inline guide/interactions/autocomplete.md:78]: This should use async/await syntax.

File: guide/interactions/autocomplete.md
Comment (adrifcastr) [inline guide/interactions/autocomplete.md:107]: Same as above

File: guide/interactions/autocomplete.md
Comment (adrifcastr) [inline guide/interactions/autocomplete.md:13]:
```suggestion
```
This feels off, what is being prepared? This entire paragraph seems redundant as the action of setting the property to true is explained below already.

**Review by adrifcastr — CHANGES_REQUESTED** [review]

File: guide/interactions/autocomplete.md
Comment (WiseDevHelper) [reply guide/interactions/autocomplete.md:58]:
```suggestion
Using <DocsLink path="class/AutocompleteInteraction?scrollTo=respond" /> you can submit an array of <DocsLink path="typedef/ApplicationCommandOptionChoice" /> objects. Passing an empty array will show "No options match your search" for the user.
```

**Review by almostSouji — CHANGES_REQUESTED** [review]
Please address or resolve reviews

**Comment by weirdG22** [comment]
Autocomplete was pushed to the main branch this morning for the Slash Command Builder. Is there a date or time, or something to push for the release of this guide?

**Comment by svemat01** [comment]
Been quite busy with school but will try to complete this asap, started working on it again today.

File: guide/interactions/autocomplete.md
Comment (svemat01) [reply guide/interactions/autocomplete.md:10]: Should `Preparing commands` be replaced with `Enabling` or `Enabling autocomplete` or should this section just be put in under `Receiving`?

File: guide/interactions/autocomplete.md
Comment (svemat01) [reply guide/interactions/autocomplete.md:21]: Should this be removed with the merge of the slash command builder which makes it now support autocomplete?

File: guide/interactions/autocomplete.md
Comment (svemat01) [reply guide/interactions/autocomplete.md]: Was resolved further down.

File: guide/interactions/autocomplete.md
Comment (vitorlops) [inline guide/interactions/autocomplete.md:59]:
```suggestion
To only display options starting with the focused value you can use the `Array#filter()` method.
```

**Review by vitorlops — COMMENTED** [review]
just a little issue with punctuation

File: guide/interactions/autocomplete.md
Comment (vitorlops) [reply guide/interactions/autocomplete.md:13]: now it does, right?

Comment (DTrombett) [reply guide/interactions/autocomplete.md:13]: Yes

**Comment by vitorlops** [comment]
@svemat01 is there any update? You didn't replied to some reviews.

**Comment by svemat01** [comment]
> @svemat01 is there any update? You didn't replied to some reviews.

I put out some questions under the comments where I didn't really know how to go forward with them and since I never really got an answer it was hard for me to continue at the time. I also got stuck in some personal issues.

However if anyone would like to hop onto a VC or a group chat somewhere where we can figure out the rest together, I would be happy to do so.

File: guide/interactions/autocomplete.md
Comment (castdrian) [reply guide/interactions/autocomplete.md:10]: merging those would probably be most senseful unless a a more appropriate header name was used

File: guide/interactions/autocomplete.md
Comment (almostSouji) [reply guide/interactions/autocomplete.md:21]: If a stable builders release support it, use builders for consistency, imo

**Comment by the-vampiire** [comment]
@svemat01 hey man thanks for putting this together. someone linked the preview for me in the djs discord and it really helped me wrap my head around how it works.

i have some notes that may be helpful to add but i can leave it to you / the maintainers to decide if and where they should be added.

> usage with subcommand options

```js
if (interaction.isAutocomplete()) {
    const command = interaction.commandName;
    const subcommand = interaction.options.getSubcommand();
    const { name: optionName, value: typedValue } = interaction.options.getFocused(true);

    console.log({
      command,
      subcommand,
      optionName,
      typedValue,
    });

    // produce results based on the details above
    // send an array of ApplicationCommandOptionChoice or empty if no matches found
    return interaction.respond([]);
  }
```

> a note about behavior after responding

- after the user selects one of the autocomplete options sent with `respond` and sends the command it will be sent as a `CommandInteraction` with the chosen option value

File: guide/interactions/autocomplete.md
Comment (BenjammingKirby) [inline guide/interactions/autocomplete.md:36]:
```suggestion
const commandData = new SlashCommandBuilder()
	.setName('autocomplete')
	.setDescription('test command to show how autocomplete should be setup')
	.setDefaultPermission(true)
	.addStringOption(opt =>
		opt
			.setName('name')
			.setDescription('Name of something')
			.setAutocomplete(true));
```

File: guide/interactions/autocomplete.md
Comment (BenjammingKirby) [inline guide/interactions/autocomplete.md:35]:
```suggestion
```js {9}
const commandData = new SlashCommandBuilder()
	.setName('autocomplete')
	.setDescription('test command to show how autocomplete should be setup')
	.setDefaultPermission(true)
	.addStringOption(opt =>
		opt
			.setName('name')
			.setDescription('Name of something')
			.setAutocomplete(true));
```

File: guide/interactions/autocomplete.md
Comment (BenjammingKirby) [inline guide/interactions/autocomplete.md:36]:
```suggestion
```js {9}
const commandData = new SlashCommandBuilder()
	.setName('autocomplete')
	.setDescription('test command to show how autocomplete should be setup')
	.setDefaultPermission(true)
	.addStringOption(opt =>
		opt
			.setName('name')
			.setDescription('Name of something')
			.setAutocomplete(true));
```

File: guide/interactions/autocomplete.md
Comment (BenjammingKirby) [inline guide/interactions/autocomplete.md:16]:
```suggestion
	.setDescription('Test command to show how autocomplete should be set up')
```

Slight grematical mistake. setup is a noun, "set up" is the verb we're probably looking for

File: guide/interactions/autocomplete.md
Comment (megatank58) [inline guide/interactions/autocomplete.md:6]:
```suggestion
This page is a follow-up to the [interactions (slash commands) page](/interactions/slash-commands.md). Please carefully read those first so that you can understand the methods used in this section.
```
The pages have been merged now so it's singular

**Review by didinele — APPROVED** [review]

**Pull request merged**
